### PR TITLE
[issue-374] implement license expression validation

### LINKED ISSUE #374: Implement license expression validation
can be done after #10 
part of #307 

Also add more tests for this

### PR DESCRIPTION
**Pull request by armintaenzertng**
fixes #374

### CODE REVIEW

File: src/spdx/validation/license_expression_validator.py [hunk 1]
```diff
@@ -11,25 +11,58 @@
...
+
+    validation_messages = []
+    license_ref_ids: List[str] = [license_ref.license_id for license_ref in document.extracted_licensing_info]
 
-    return error_messages
+    for non_spdx_token in get_spdx_licensing().validate(license_expression).invalid_symbols:
+        if non_spdx_token not in license_ref_ids:
+            validation_messages.append(
+                ValidationMessage(
+                    f"Unrecognized license reference: {non_spdx_token}. license_expression must only use IDs from the license list or extracted licensing info, but is: {license_expression}",
```
Comment (meretp) [inline src/spdx/validation/license_expression_validator.py:51]:
```suggestion
                    f"Unrecognized license reference: {non_spdx_token}. License_expression must only use IDs from the license list or extracted licensing info, but is: {license_expression}",
```

File: tests/spdx/validation/test_license_expression_validator.py [hunk 2]
```diff
@@ -10,14 +10,94 @@
...
+                           f'A plain license symbol cannot be used as an exception in a "WITH symbol" statement. for token: "{FIXTURE_LICENSE_ID}" at position: 39. for license_expression: GPL-2.0-or-later AND {FIXTURE_LICENSE_ID} WITH {FIXTURE_LICENSE_ID}'),
+                          (f"GPL-2.0-or-later with MIT and {FIXTURE_LICENSE_ID} with GPL-2.0-or-later",
+                           f'A plain license symbol cannot be used as an exception in a "WITH symbol" statement. for token: "MIT" at position: 22. for license_expression: GPL-2.0-or-later WITH MIT AND {FIXTURE_LICENSE_ID} WITH GPL-2.0-or-later'),
+                          ("389-exception with 389-exception",
+                           'A license exception symbol can only be used as an exception in a "WITH exception" statement. for token: "389-exception". for license_expression: 389-exception WITH 389-exception'),
+                          ("389-exception with MIT",
+                           'A license exception symbol can only be used as an exception in a "WITH exception" statement. for token: "389-exception". for license_expression: 389-exception WITH MIT'),
```
Comment (meretp) [inline tests/spdx/validation/test_license_expression_validator.py:91]: The validation provided by the `license_expression` package seems to return after the first error found which is not consistent with our approach. That's unfortune but probably not changeable, only to mention it here.

File: src/spdx/validation/license_expression_validator.py [hunk 3]
```diff
@@ -11,25 +11,58 @@
...
+        get_spdx_licensing().parse(str(license_expression), validate=True, strict=True)
+    except ExpressionParseError as err:
+        validation_messages.append(
+            ValidationMessage(
+                f"{err}. for license_expression: {license_expression}",
+                context)
+        )
+    except ExpressionError:
+        # This happens for invalid symbols but the error provides only a string of these. On the other hand,
+        # get_spdx_licensing().validate() gives an actual list of invalid symbols, so this is handled above.
```
Comment (meretp) [inline src/spdx/validation/license_expression_validator.py:65]:
```suggestion
        # This error is raised for invalid symbols within the license_expression but it provides only a string of these. On the other hand,
        # get_spdx_licensing().validate() gives an actual list of invalid symbols, so this is handled above.
```
At first I was very confused about this comment, but with the previous discussion and the discussion in the [issue](https://github.com/nexB/license-expression/issues/82) in mind it got clearer. Although I'm not a fan of comments in general, maybe you could add one for the `ExpressionParseError` that it is used to catch falsy exceptions.

File: src/spdx/validation/license_expression_validator.py (hunk 1 above)
Comment (meretp) [reply src/spdx/validation/license_expression_validator.py:51]: Seeing that the error message from the `ExpressionParseError` also uses lower case after a `.` you can ignore this remark to keep it "somehow" consistent, but I don't really like this.

**Review by meretp — APPROVED** [review]
Thanks for adding this! Only a few nitpicking remarks from my side that are not merge-blocking.

File: src/spdx/validation/license_expression_validator.py (hunk 1 above)
Comment (armintaenzertng) [reply src/spdx/validation/license_expression_validator.py:51]: yeah, not ideal, but it should work nevertheless ;)

File: src/spdx/validation/license_expression_validator.py [hunk 4]
```diff
@@ -11,25 +11,58 @@
...
+
+    return validation_messages
+
+
+def validate_license_expression(license_expression: Optional[
+    Union[LicenseExpression, SpdxNoAssertion, SpdxNone]], document: Document, parent_id: str, context: ValidationContext = None) -> List[ValidationMessage]:
+    if license_expression in [SpdxNoAssertion(), SpdxNone(), None]:
+        return []
+
+    if context is None:
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply src/spdx/validation/license_expression_validator.py:41]: done

File: tests/spdx/validation/test_license_expression_validator.py (hunk 2 above)
Comment (armintaenzertng) [reply tests/spdx/validation/test_license_expression_validator.py:91]: Yes, this is also not ideal, but it really is only a workaround until maybe `get_spdx_licensing().validate()` lists the full list of errors.

File: src/spdx/validation/license_expression_validator.py (hunk 3 above)
Comment (armintaenzertng) [reply src/spdx/validation/license_expression_validator.py:65]: done

File: src/spdx/validation/validation_message.py [hunk 5]
```diff
@@ -25,7 +26,6 @@ class SpdxElementType(Enum):
     PACKAGE = auto()
     FILE = auto()
     SNIPPET = auto()
-    LICENSE = auto()
```
Comment (meretp) [inline src/spdx/validation/validation_message.py:28]: Good catch!

**Review by meretp — APPROVED** [review]
Thanks!

**Pull request merged**